server/node.go

```diff
@@ -171,11 +172,52 @@ func NewNode(db *client.KV, gossip *gossip.Gossip, storeConfig storage.StoreConf
 // initDescriptor initializes the node descriptor with the server
 // address and the node attributes.
 func (n *Node) initDescriptor(addr net.Addr, attrs proto.Attributes) {
-	n.Descriptor = storage.NodeDescriptor{
-		// NodeID is set after invocation of start()
-		Address: addr,
-		Attrs:   attrs,
+	n.Descriptor.Address = addr
+	n.Descriptor.Attrs = attrs
+}
+
+// initNodeID updates the internal NodeDescriptor with the given ID. If zero is
+// supplied, a new NodeID is allocated with the first invocation. For all other
+// values, the supplied ID is stored into the descriptor (unless one has been
+// set previously, in which case a fatal error occurs).
+//
+// Upon setting a new NodeID, the descriptor is gossiped and the NodeID is
+// stored into the gossip instance.
+func (n *Node) initNodeID(id proto.NodeID) {
+	if id < 0 {
+		log.Fatalf("NodeID must not be negative")
+	}
+
+	if o := n.Descriptor.NodeID; o > 0 {
+		if id == 0 {
+			return
+		}
+		log.Fatalf("cannot initialize NodeID to %d, already have %d", id, o)
+	}
+	var err error
+	if id == 0 {
+		id, err = allocateNodeID(n.db)
+		if err != nil {
+			log.Fatal(err)
+		}
+		if id == 0 {
+			log.Fatal("new node allocated illegal ID 0")
+		}
+
+		log.Infof("new node allocated ID %d", n.Descriptor.NodeID)
 	}
+
+	n.Descriptor.NodeID = id
+
+	nodeIDKey := gossip.MakeNodeIDKey(n.Descriptor.NodeID)
+	if err = n.gossip.AddInfo(nodeIDKey, &n.Descriptor, ttlNodeIDGossip); err != nil {
+		log.Fatalf("couldn't gossip address for node %d: %v", n.Descriptor.NodeID, err)
+	}
+	// Setting the NodeID as the name of the gossip instance allows the
+	// DistSender to look up the own node's descriptor from the gossip network.
+	// The field is touched only once, so the atomic update isn't necessary but
+	// it feels better.
```

Comment:
We should have a better justification than this when doing things related to concurrency. If we do an atomic store here we probably need to do an atomic load wherever we read it. If we can, I'd rather do a simple assignment here since this is only done once at startup. If we need to use atomic ops then we should unexport the NodeID field and add a SetNodeID method to Gossip so the access to this field can be contained in one place.